ui_test/src/lib.rs

```diff
@@ -293,50 +316,111 @@ fn check_test_result(
+        // in the messages.
+        if let Some(i) = diagnostics
+            .messages_from_unknown_file_or_line
+            .iter()
+            .position(|msg| msg.message.contains(error_pattern))
+        {
+            diagnostics.messages_from_unknown_file_or_line.remove(i);
+            found = true;
+        }
+
+        // if nothing was found, check the ones inside our file. We permit this because some tests may have
+        // flaky line numbers for their messages.
+        if !found {
+            for line in &mut diagnostics.messages {
+                if let Some(i) = line.iter().position(|msg| msg.message.contains(error_pattern)) {
+                    line.remove(i);
+                    found = true;
+                    break;
+                }
+            }
+        }
+
+        if !found {
             errors.push(Error::PatternNotFound {
-                stderr: unnormalized_stderr.to_string(),
+                stderr: diagnostics.rendered.to_string(),
                 pattern: error_pattern.to_string(),
                 definition_line,
             });
         }
-        found_annotation = true;
     }
-    for &ErrorMatch { ref matched, revision: ref rev, definition_line } in &comments.error_matches {
-        // FIXME: check that the error happens on the marked line
-
+    for &ErrorMatch { ref matched, revision: ref rev, definition_line, line, level } in
+        &comments.error_matches
+    {
         if let Some(rev) = rev {
             if rev != revision {
                 continue;
             }
         }
+        if let Some(level) = level {
+            // If we found a diagnostic with a level annotation, make sure that all
+            // diagnostics of that level have annotations, even if we don't end up finding a matching diagnostic
+            // for this pattern.
+            highest_annotation_level = std::cmp::max(highest_annotation_level, level);
+        }
 
-        if !unnormalized_stderr.contains(matched) {
-            errors.push(Error::PatternNotFound {
-                stderr: unnormalized_stderr.to_string(),
-                pattern: matched.to_string(),
-                definition_line,
+        if let Some(msgs) = diagnostics.messages.get_mut(line) {
+            let found = msgs.iter().position(|msg| {
+                msg.message.contains(matched)
+                    // in case there is no level on the annotation, match any level.
+                    && level.map_or(true, |level| {
+                        msg.level == level
+                    })
             });
+            if let Some(found) = found {
+                let msg = msgs.remove(found);
+                // If we found a diagnostic, even without having any level annotation, make sure that all
+                // other diagnostics of that level also have annotations.
+                highest_annotation_level = std::cmp::max(highest_annotation_level, msg.level);
```

Comment:
highest_annotation_level should only be raised when the comment specifies a level.

When I use `//~NOTE` to match a note, I do *not* have to also match all other notes. I have crucially relied on this compiletest feature.^^

Reply:
seems footgunny. I'd rather add more magic sigils

Reply:
Sorry I had a fatal typo in my previous comment. It should have been

> When I use `//~ some text` to match a note, I do not have to also match all other notes. I have crucially relied on this compiletest feature.^^

So, I see `//~LEVEL` as a request "please make sure we match all things at that level". That request is implicit, sure, but at least I am stating the level explicitly. To do the same when a level is not even stated seems waay to implicit to me.

Reply:
Ok, I guess that makes sense. What irked me in the rustc test suite though is that those will also match an error-level diagnostic. I think we should just always require `ERROR` in annotations, to distinguish it from the "just check that the message is here, don't start checking all other things at this level" case

Reply:
I would be fine with saying that patterns without level cannot match ERROR. (As long as that is documented.)

Reply:
Done